### ISSUE TITLE
[Help]: About processing big data 

### ISSUE THREAD

**Issue by Twcumtb** [gmtsar/gmtsar#413]
I have 139 Sentinel-1 SAR scenes and formed 380 interferograms. The dimension of each interferogram is 8556*7550. I used SBAS_parallel to run the time series analysis.  I know that it is computationally expensive with such big data when dosing the SBAS with GMTSAR. My RAM is already 128GB. But I still can't handle this data by the script SBAS_parallel. It just killed the processing and the output message says "Required Memory Usage is 341.9 GB".  Sometimes I can not afford such big RAM (341GB) on my computer. How can I process such kind of big data with my current RAM (128 GB)?

**Comment by SteffanDavies** [comment]
Use the -mmap option.

**Comment by Twcumtb** [comment]
> 

Hi SteffanDavies,
   Thank you very much for your reply. And I tried to use -mmap option in the sbas_parallel command like "sbas_parallel intf.tab scene.tab 380 139 8556 7550 -atm 3 -smooth 3 -wavelength 0.0554658 -incidence 39.21622 -range 845621.62 -rms -dem -mmap". But I still can not successfully processed my data with a large coverage and a large stack. That it has got stuck for a week already in the first iteration as shown in the figure. How to handle this issue? 

![WechatIMG900](https://user-images.githubusercontent.com/43894333/191446759-66bc103e-0994-4aaa-bb56-a7a443561acf.jpeg)

- [ ]

**Comment by SteffanDavies** [comment]
Did -mmap solve the ram constraints?

**Comment by Twcumtb** [comment]
> 

Yes. The RAM constraint was solved. The Required Memory Usage is 67.631846 GB ...
The output message is as follows:

sbas_parallel intf.tab scene.tab 380 139 8556 7550 -atm 3 -smooth 3 -wavelength 0.0554658 -incidence 39.21622 -range 845621.62 -rms -dem -mmap

compute RMS misfit
compute DEM error
mmap disk space for less use of memory

setting smoothing to   3.000 
setting radar wavelength to   0.055 m 
setting radar incidence angle to  39.216 degree 
setting range to 845621.620 m 
Required Memory Usage is 67.631846 GB ...
Memory Allocation Successful...
read table file ...
number of SAR scenes is 139 
number of interferograms is 380 
read phase and correlation grids ...
3.000000 0.000424 0.000000 22.971975


Applying atmospheric correction by common point stacking...

**Comment by SteffanDavies** [comment]
How many CPUs are running in parallel?

**Comment by Twcumtb** [comment]
> How many CPUs are running in parallel?

Do you mean the threads? I set the variable OMP_NUM_THREADS to 10 by command "export OMP_NUM_THREADS=10" in the terminal, and then ran the sbas_parallel.

**Comment by mobigroup** [comment]
@Twcumtb you’d try PyGMTSAR which uses lazy computations. I tested SBAS calculation on a stack of ~800 partial interferograms on 16GB RAM laptop without any problem. Use detrending instead of GMTSAR SBAS atmospheric correction. See the link: https://github.com/mobigroup/gmtsar There are some live Google Colab examples including a SBAS example with detrending.

**Comment by mobigroup** [comment]
Tested on 792 single-subswath interferograms (3208x6094 grid) on iMac Apple Silicon (16 GB RAM and 8 cores) in Python3 Jupyter notebook: 
<img width="657" alt="image" src="https://user-images.githubusercontent.com/7342379/192128712-1c21d5d9-27d6-4474-aa69-ed32ac753c2f.png">

**Comment by Luvina0503** [comment]
@Twcumtb 
Can you send me your sbas_parallel text?

**Comment by Twcumtb** [comment]
> 

On my computer, I don't need to add any extra commands.  I did not have any idea about the problem of this -mmp could not recognized.

**Comment by Luvina0503** [comment]
The sbas-mmap issue has been resolved. Thanks for your patience

**Comment by Luvina0503** [comment]
> @Twcumtb you’d try PyGMTSAR which uses lazy computations. I tested SBAS calculation on a stack of ~800 partial interferograms on 16GB RAM laptop without any problem. Use detrending instead of GMTSAR SBAS atmospheric correction. See the link: https://github.com/mobigroup/gmtsar There are some live Google Colab examples including a SBAS example with detrending.

@mobigroup @Twcumtb 
60 views sentinel data, 207 interference pairs, 16 gigabytes of computer memory
How do I use PyGMTSAR? Is there a detailed study course? I am a beginner GMTSAR and have only now learnt to tsar a set of processes. I currently work on the sbas_parallel-mmap command, but the progress is too slow. I'm very lucky to have read this reply, but I don't know how to use PyGMTSAR very well. Hope to receive your reply soon, thank you

**Comment by Luvina0503** [comment]
@mobigroup 
Can PyGMTSAR be installed on top of GMTSAR? Or do you need to start over?

**Comment by mobigroup** [comment]
@Luvina0503 There are a lot of ways how to use PyGMTSAR: on Google Colab, in Docker containers or just have a local installation. See the GitHub https://github.com/mobigroup/gmtsar and DockerHub https://hub.docker.com/u/mobigroup and the project documentation https://mobigroup.github.io/gmtsar/ for the details.

**Comment by Luvina0503** [comment]
@mobigroup 
But my google browser can't open Google Colab

**Comment by mobigroup** [comment]
@Luvina0503 Hmm, I'm confused if your computer obviously have Internet connection and can open GitHub but not Google Colab. Could you explain the situation?..

**Comment by Luvina0503** [comment]
@mobigroup 
Now I can go into the colab. But I have another problem. I've already done sbas with GMTSAR, should I go on to do sbas with colab? Or all of my steps to start over? Could you please share your code? I was hoping to use the Sentinel data for subsidence analysis, but I couldn't find any cases.

**Comment by Luvina0503** [comment]
![FC}%M}VH`GAZG${HFZ{8EF](https://user-images.githubusercontent.com/114629377/205528513-8356581a-408e-43fd-bcd2-052fe2c7f7f5.png)

**Comment by Luvina0503** [comment]
![image](https://user-images.githubusercontent.com/114629377/205529950-f63dc6d9-b49c-473d-9ffd-4ae15077f9f0.png)
What does this window's.py file do?

**Comment by mobigroup** [comment]
> Could you please share your code? 

The library code is shared on GitHub and PyPI. The examples notebooks shared on GitHub, Google Colab, DockerHub.

> I was hoping to use the Sentinel data for subsidence analysis

The example notebooks available on GitHub https://github.com/mobigroup/YamchiDam and in the Docker images https://hub.docker.com/r/mobigroup/pygmtsar-large

> What does this window's.py file do?

https://docs.python.org/3/distutils/setupscript.html

You’d run the small SBAS example on Google Colab and the large example in Docker containers to be more confident with the processing. After that just placed upyour own scenes into the data directory and run the selected notebook.

**Comment by Luvina0503** [comment]
60 views sentinel data, 207 interference pairs

Will the time of sbas analysis on docker be a little shorter? I have been running sbas-mmap on VM for 3 days and it is not finished yet.

**Comment by Luvina0503** [comment]
@mobigroup I installed docker on Ubuntu. It is still best to install docker desktop on Windows

**Comment by mobigroup** [comment]
@Luvina0503 Docker Desktop is just a user-friendly interface for docker so you can use it too. As soon as you have docker installed use DockerHub PyGMTSAR images (or build them yourself).

**Comment by cxinlong** [comment]
> > 
> 
> Hi SteffanDavies, Thank you very much for your reply. And I tried to use -mmap option in the sbas_parallel command like "sbas_parallel intf.tab scene.tab 380 139 8556 7550 -atm 3 -smooth 3 -wavelength 0.0554658 -incidence 39.21622 -range 845621.62 -rms -dem -mmap". But I still can not successfully processed my data with a large coverage and a large stack. That it has got stuck for a week already in the first iteration as shown in the figure. How to handle this issue?
> 
> ![WechatIMG900](https://user-images.githubusercontent.com/43894333/191446759-66bc103e-0994-4aaa-bb56-a7a443561acf.jpeg)
> 
> * [ ]

Hello, may I ask if this problem has been solved？ @Twcumtb

**Comment by mobigroup** [comment]
@cxinlong There is Docker image to process https://hub.docker.com/r/mobigroup/pygmtsar-large to process hundreds of large interferograms ~20000 x 20000 pixels size on 16 GB RAM host. It works well even on Apple Air laptop. The discussed interferograms are much smaller (“8556*7550”). What do you need more?

**Comment by cxinlong** [comment]
> 

Thank you for your answer. How should I operate? Do I need to install Docker? I use ubuntu system.

**Comment by cxinlong** [comment]
> @cxinlong There is Docker image to process https://hub.docker.com/r/mobigroup/pygmtsar-large to process hundreds of large interferograms ~20000 x 20000 pixels size on 16 GB RAM host. It works well even on Apple Air laptop. The discussed interferograms are much smaller (“8556*7550”). What do you need more?

Thank you for your answer. How should I operate? Do I need to install Docker? I use ubuntu system. @mobigroup

**Comment by cxinlong** [comment]
@mobigroup Now I have the same problem as the author of this topic, sbas command has been running for a long time, but still can't get the result. I don't know how to deal with this situation. So I wonder how the author deals with it? Then, according to the Docker you said, what exactly is it used for? Can you solve the problem of not getting the result? Because I can't open the link you gave me, sorry.

**Comment by mobigroup** [comment]
@cxinlong The Docker image includes GMTSAR binaries and PyGMTSAR Python tools. PyGMTSAR provides different SBAS processing pipeline where we use separate detrending operation (if needed) and least-squares processing for all the SBAS pairs. This way the SBAS processing is optimized and the problem is resolved. Also, PyGMTSAR allows SNAPHU tiling unwrapping using multicore unwrapping for every separate interferogram and utilizing less memory.

**Comment by Twcumtb** [comment]
> @mobigroup Now I have the same problem as the author of this topic, sbas command has been running for a long time, but still can't get the result. I don't know how to deal with this situation. So I wonder how the author deals with it? Then, according to the Docker you said, what exactly is it used for? Can you solve the problem of not getting the result? Because I can't open the link you gave me, sorry.

I tried it by using -mmap, but the problem did not solve. I cut the interferograms into several patches and processed them separately by sbas. And paste the velocity results together.

**Comment by cxinlong** [comment]
Can you tell me how to cut the interferogram and paste the velocity results？How does it work？发自我的手机-------- 原始邮件 --------发件人： Twcumtb ***@***.***>日期： 2023年2月16日周四 17:21收件人： gmtsar/gmtsar ***@***.***>抄送： cxinlong ***@***.***>, Mention ***@***.***>主    题： Re: [gmtsar/gmtsar] [Help]: About processing big data  (Issue #413)

@mobigroup Now I have the same problem as the author of this topic, sbas command has been running for a long time, but still can't get the result. I don't know how to deal with this situation. So I wonder how the author deals with it? Then, according to the Docker you said, what exactly is it used for? Can you solve the problem of not getting the result? Because I can't open the link you gave me, sorry.

I tried it by using -mmap, but the problem did not solve. I cut the interferograms into several patches and processed them separately by sbas. And paste the velocity results together.

**Comment by mobigroup** [comment]
@cxinlong

> I tried it by using -mmap, but the problem did not solve.

GMTSAR SBAS merges two-stage processing into one step: atmospheric correction and the weighted least-squares pixel-wise movements calculations. It does not work for large stacks of large grids. GMTSAR developers apply decimation to process low-resolution interferograms and use hosts having like to 512 GB RAM and wait for weeks for the results. Actually, that is the reality for all the interferometry packages but I want to do the same and much more on my Apple Air and that was the reason to create PyGMTSAR.

Cutting interferograms you have a problem with inconsistent unwrapping.

**Comment by cxinlong** [comment]
So you mean that GMTSAR can run SBAS and deal with low resolution interferograms, It just takes a few weeks?Would you have gotten the same result if you had used pyGMTSAR？Does pyGMTSAR need the same file as GMTSAR to run SBAS?And Do you have a manual for pyGMTSAR to process data? @mobigroup

**Comment by cxinlong** [comment]
Can you tell me how to cut the interferogram and paste the velosity results?How does is work? @Twcumtb

**Comment by mobigroup** [comment]
@cxinlong I've send you the links above which are able to run in just one click. Also, see the manual and explanation links above.

**Comment by cxinlong** [comment]
I can't open this link: https://medium.com/@pechnikov/pygmtsar-is-next-generation-interferometric-synthetic-aperture-radar-insar-software-for-ev eryone-7e2607d103cf. Do you know the answer to my first question? Just run the SBAS with GMTSAR for enough time could get the results. @mobigroup

**Comment by mobigroup** [comment]
@cxinlong Your link is not correct because you've added space inside it. 

> Just run the SBAS with GMTSAR for enough time could get the results.

You'd try. Just wait about a week to be sure if it works.

**Comment by cxinlong** [comment]
Sorry, I have tried many times but I can't open this link. I don't know where there is an extra space. Can you show me what I should do now? Need to install python first? And install pyGMTSAR? @mobigroup

**Comment by cxinlong** [comment]
I don't know why,but it still won't open. @mobigroup

**Comment by mobigroup** [comment]
@cxinlong The same publication is available here: https://habr.com/ru/post/717122/

**Comment by cxinlong** [comment]
ok,it can be opened.Thank you very much. @mobigroup

**Comment by behshidkhodaei** [comment]
> The sbas-mmap issue has been resolved. Thanks for your patience

Can I ask how you could fix that issue with sbas_parallel .... -mmap as when I am using that option at the end of command, I get a message which says "option not recognized"?

**Comment by cxinlong** [comment]
> Did you try turning on the -mmap option? If you turn off -atm, you could cut things into smaller pieces, run sbas and paste them together afterwards.

Hello.If first cut into several pieces for sbas processing, and then paste. The benchmark is not the same when each block is processed. If you paste it directly, will there be deformation value jump at the joint?

**Comment by Xiaohua-Eric-Xu** [comment]
Did you turn off -atm? Could you show an example? Also you should not perform anything after the cutting.

**Comment by SteffanDavies** [comment]
@mobigroup can I run PyGMTSAR SBAS in ./SBAS folder on dataset prepared for gmtsar SBAS without reprocessing?

**Comment by mobigroup** [comment]
@SteffanDavies I think that’s possible while some coding required. You need to rename input files to PyGMTSAR naming convention and rename the output files to GMTSAR names. But what is the reason to do it? GMTSAR processing is slower because it uses single CPU core for DEM processing, repeats slow geocoding for all the rasters, and so on.

**Comment by SteffanDavies** [comment]
> @SteffanDavies I think that’s possible while some coding required. You need to rename input files to PyGMTSAR naming convention and rename the output files to GMTSAR names. But what is the reason to do it? GMTSAR processing is slower because it uses single CPU core for DEM processing, repeats slow geocoding for all the rasters, and so on.

My implementation of GMTSAR is highly parallel using bash/docker/python and cloud however I am reaching the limits of what is reasonable in both cost and efficiency, and am thinking of moving to PyGMTSAR very soon. However I have very little experience with it and a concise tutorial would be very good for your project (I'm not talking about notebooks). Especially considering I have had a couple of bugs/issues (some have been resolved as you know) See how many users on here refer to GMTSAR Sentinel-1 recipe and UNAVCO course to start understanding GMTSAR - many.

My current dataset has almost 1000 unwrapped interferograms  and has been processing for a couple of weeks however GMTSAR requires 700gb tam without mmap option and writing to disk takes forever as I am limited in this instance, hence my previous question.

Thanks

**Comment by mobigroup** [comment]
Ok, understand. I’ll try to do the SBAS processing on the GMTSAR results. PyGMTSAR geocoding is not possible for this case so you would use GMTSAR scripts for the SBAS output.
And about the Sentinel-1 recipe. How do you see it when the ready to use notebooks provided and PyGMTSAR documentation is available on Github: https://mobigroup.github.io/gmtsar/ The notebooks can download the input scenes and orbits and DEM so the pipeline looks straightforward, I hope. The output vertical and west-east displacements for two orbits should be just summarized. Maybe I’ve missed some points but I don’t know what ones. Do you have some suggestions? 
Actually, I think about the complete processing on Python without GMTSAR tools to allow parallel processing for a single interferogram to archive 8-128 times faster processing on modern multicore computers and limit RAM consumption but it looks outside of your tasks. Multi scale unwrapping is interesting too because it would be much faster than SNAPHU processing but it’s complicated and I have no enough free time to work on it. Anyway, the both features are not valuable for GMTSAR users as I see.

**Comment by madeoka** [comment]
Hello Mr @mobigroup , Thank you for your awesome work, I have a question can I share the local disk (for example E: drive) to the pygmtsar container?
I try to do that, but until now, I cannot do it, do you have any advice for me?
I am so sorry I full beginner in using docker.

**Comment by mobigroup** [comment]
@madeoka Sure, that's possible for GMTSAR and PyGMTSAR Docker images. You can see the already mounted directories in Docker Desktop settings and add new ones as needed:

![Docker Desktop Settings](https://github.com/gmtsar/gmtsar/assets/7342379/07cbdfcc-2a05-46e0-91ed-9f63f8fe0e98)

Feel free to use this approach for read-only Sentinel-1 scenes, when all your work files are stored inside a container. You can find more details in the Docker documentation: [Docker File Sharing on macOS](https://docs.docker.com/desktop/settings/mac/#file-sharing).

**Comment by madeoka** [comment]
@mobigroup I am so sorry Mr, I forgot to mention, I work using windows10 pro.
But the docker using WSL2 Ubuntu backend, no has tab "File sharing".
Do you ever get this problem?
![image](https://github.com/gmtsar/gmtsar/assets/16494453/582354f4-c346-49bb-b4ef-7bbbcc5e35b4)

**Comment by mobigroup** [comment]
How about https://medium.com/analytics-vidhya/sharing-windows-folders-for-containers-4539f7ed2acf

**Comment by madeoka** [comment]
Thank you Mr @mobigroup , I try it and can solved sharing folders.
Now, my problem is how to integrated with your container and how to change the container disk to another disk, because the docker using much my C disk.
I still finding this solver.

**Comment by mobigroup** [comment]
You might configure the default Docker root folder (I think it is C:\ProgramData\DockerDesktop on Windows).

**Comment by madeoka** [comment]
Hi Mr @mobigroup finally I can setting, I am using WSL2 as backend. So, I can changing the docker root folder in another disk. For data, it cannot be handle like you mention above, but I try to using "import" function in Files for each container. Now, I will try to processing my data.
Thank you for your guiding, if any problem occur, please your help.
Thank you in advance.
![image](https://github.com/gmtsar/gmtsar/assets/16494453/8632168d-64e0-4d7a-8ef7-d0b24ecb0560)
